core/schema/env.graphqls

```diff
@@ -0,0 +1,98 @@
+extend type Query {
+  "Load a environment from ID."
+  environment(id: EnvironmentID): Environment!
+
+  "Load a environment check from ID."
+  environmentCheck(id: CheckID): Check!
+
+  "Return the current environment being executed in."
+  currentEnvironment: Environment!
+}
+
+"A unique environment identifier."
+scalar EnvironmentID
+
+"A unique environment check identifier."
+scalar CheckID
+
+"A unique environment check result identifier."
+scalar CheckResultID
+
+"A group of Dagger entrypoints that can be queried and/or invoked."
+type Environment {
+  "A unique identifier for this environment."
+  id: EnvironmentID!
+
+  "Initialize this environment from the given directory and config path"
+  load(source: DirectoryID!, configPath: String!): Environment!
+
+  "Name of the environment"
+  name: String!
+
+  "This environment plus the given check"
+  withCheck(id: CheckID!): Environment!
+
+  "The list of checks in this environment"
+  checks: [Check!]
+
+  "The check in this environment with the given name, if any"
+  check(name: String!): Check
+}
+
+"An entrypoint for tests, lints or anything that can pass/fail."
+type Check {
+  "A unique identifier for this check."
+  id: CheckID!
+
+  "The name of the check."
+  name: String!
+
+  "This check with the given name"
+  withName(name: String!): Check!
+
+  "Documentation for this check."
+  description: String
+
+  "This check with the given description"
+  withDescription(description: String!): Check!
+
+  "The subchecks of this check."
+  subchecks: [Check!]
+
+  "This check with the given subcheck"
+  withSubcheck(id: CheckID!): Check!
+
+  """
+  This check with the given container used to determine the check's result.
+  If set, the container will be executed and the check result will be set to
+  success if the container exits with a zero exit code, failure otherwise.
+  """
+  withContainer(id: ContainerID!): Check!
```

Comment:
This API was suggested by @vito while he was implementing dynamic checks derived from go tests: https://github.com/dagger/dagger/blob/2763c2241e88a3282622652ff7309550982b473f/universe/go/main.go#L179

It's extremely convenient for those sorts of dynamic use cases, so I think we should start with it in the API. If we can find a better way of supporting dynamic checks that only relies on entrypoint env code, then it may become slightly redundant and we can consider whether we still need it.